### DISCUSSION TITLE
Support for decorators stage 3 proposal with `accessor`

### DISCUSSION THREAD

**Discussion by WearyMonkey** [mobxjs/mobx#3373] [General] [answered]
The decorators spec has progressed to stage 3 and has introduced the new `accessor` keyword and the ability to capture initial field values. On the surface these features seem to resolve the mobx decorator incompatibility with the previous static decorator proposal.

Is there any plan to introduce support for decorating `accessor` fields without the need for `makeObservable`  in future mobx versions? Or are you currently waiting for further progress of the proposal, e.g. stage 4 or TypeScript support?

Context: we currently have a very large code base on mobx 5. We would like to upgrade to mobx 6 but consider it a risk due to the required `makeObservable`, which if missing will cause subtle runtime issues.

Ideally, if mobx were to support the accessor keyword in a future version, we would be able to leapfrog mobx 6 to the next version.

https://github.com/tc39/proposal-decorators#class-auto-accessors
https://github.com/tc39/proposal-decorators#class-fields

Links to related issues:
https://github.com/mobxjs/mobx/issues/1928
https://github.com/tc39/proposal-decorators/issues/267
https://github.com/mobxjs/mobx/pull/1732
https://github.com/microsoft/TypeScript/issues/48686 (Some discussion in TS about stage 3 support)

Thank you for all your awesome work!

**Comment by mweststrate** [comment, marked as answer]
MobX 6 deprecated decorators exactly because we want to be able to switch to the standard proposal in the future. I've been involved designing the proposal, and we'll pick it up further once a transpilers are available. 

> Context: we currently have a very large code base on mobx 5. We would like to upgrade to mobx 6 but consider it a risk due to the required makeObservable, which if missing will cause subtle runtime issues.

Please note that a codemod is available

**Reply by WearyMonkey** [reply]
Do you think `makeObservable` will be required after switching to the new standard? Or is that unknown at this point?

**Reply by phiresky** [reply]
> we'll pick it up further once a transpilers are available.

The new decorators support has finally been merged to TypeScript! It's part of `typescript@5.0.0-dev.20230120` and enabled without any config flags (by removing `experimentalDecorators: true` from the tsconfig).

**Comment by mweststrate** [comment]
In principle no, it's a stopgap. But please note that proposals move
slowly, this one has taken 7 years at least so far.

**Reply by phiresky** [reply]
Just wanted to mention that the new (stage 3) decorators support is about to land in TypeScript Nightly with this PR: https://github.com/microsoft/TypeScript/pull/50820 and is slated to be released in TypeScript 5.0 (the next version): https://github.com/microsoft/TypeScript/issues/51362

Has the time come to revisit this? :)

**Comment by gilamran** [comment]
Typescript 5 will support stage 3 decorators, do you plan supporting it?

**Reply by mweststrate** [reply]
yes :) Practically not sure yet when.

**Comment by Matchlighter** [comment]
@mweststrate Can I ask where 2022.3/stage 3 decorators place on your timeline/priorities and whether you've outlined any plans/requirements/test cases? This would be a big feature for me both at home and work, so I want to do a PR if you (or someone else on the MobX team) are available to jump on reviews and bounce with me on iterations.

**Reply by mweststrate** [reply]
@Matchlighter I was planning to look into that ~end march / april. I didn't outline any plans or test cases, but in principle we have all the test cases as we still have tests that test decorator support, where the important bit is that they should still work even when `makeObservable(this)` is not called in the constructor (that should be a no-op). But happy to review PRs if you are willing to look into it! Upfront I don't give merge guarantees pending on code quality, implementation alignment etc etc as a disclaimer. That being said, happy to receive the help, and happy to give pointers where needed!

**Reply by Matchlighter** [reply]
Cool cool.
Digging through the code base, I'm seeing a lot of work around "Annotations" and `make_`/`extend_`. The first approach I could see would be to just tie into these and (effectively) use an `addInitializer` (from the 2022.3 decorator spec) callback to essentially call `makeObservable(this, { decoratee: observable })`.
However, while not totally incompatible with 2022.3 decorators, it poses some large limitations, particularly when wanting to chain decorators. For example, given:
```typescript
class Thing {
  @custom_decorator
  @observable
  accessor my_field = 4
}
```
Using the current annotation system (either with `makeObservable` or piggy backing off it with `addInitializer`) will end up re-`defineProperty`ing the `my_field` property when it later `make_`/`extend_`/applies the annotations. Due to that, my implementation plan would be to circumvent the annotation system and use the standard decorator API for modifying/wrapping the decoratee, using `addInitializer` if/where it's necessary. My thought is to inject a `if (typeof arg2 == 'object' && typeof arg2.kind == 'string')` clause into the `api`-domain functions (eg above line 98 of https://github.com/mobxjs/mobx/blob/faf075cc70a16a86443b24c8dd120516bac016ed/packages/mobx/src/api/observable.ts#L95-L100) to catch the decorator use case and apply expected get/set/call hooks.

Stated more shortly - I plan on keeping the annotation API for backwards compatibility, but bypassing it when using 2022.3 decorator syntax. (Unless you want to go all out and drop another major version, but I'm rarely able to sell that proposition :P)

I foresee a possible breaking change in that interceptors as they're used in the many `if (hasInterceptors(this))` blocks in https://github.com/mobxjs/mobx/blob/main/packages/mobx/src/types/observableobject.ts may not be applicable - is there a time that interceptors would be added to an object before it calls `makeObservable()`?

This approach does mean that a lot of the existing code (the complete annotation system) will be unused by the decorator code, and thus there may be some (though I think minor) duplication.

What are your general thoughts/👍-get-coding/👎-something's-off given that schload of info?

**Reply by Matchlighter** [reply]
Well, I got bored/excited and jumped in either way! I've found a way to extend the `Annotation` interface to handle decorators. Still having to duplicate some small bits of code, but this at least keeps everything neatly in one place.
```typescript
export type Annotation = {
    annotationType_: string
    make_(
        adm: ObservableObjectAdministration,
        key: PropertyKey,
        descriptor: PropertyDescriptor,
        source: object
    ): MakeResult
    extend_(
        adm: ObservableObjectAdministration,
        key: PropertyKey,
        descriptor: PropertyDescriptor,
        proxyTrap: boolean
    ): boolean | null
    decorate_20223_( // <<-- Added
        value: any,
        context: DecoratorContext
    )
    options_?: any
}
```
I think I've got most of the primary cases addressed, but am calling it a night. I'm hoping to figure out the test suite tomorrow and ideally have an early PR up as well.

Current complete code at https://github.com/Matchlighter/mobx/tree/decorators2022

**Comment by belaczek** [comment]
Technical question: Once the stage 3 decorators support is implemented in mobx, is it possible that the requirement for `useDefineForClassFields: true` will be removed? Or is that expected to stay?

**Reply by Matchlighter** [reply]
I've got a PR open for Stage 3 Decs. My approach was to go the way ES was going, so I wrote and tested things with `useDefineForClassFields: true`. That said, I don't think I did anything that would actually require `useDefineForClassFields: true`, so it's likely up to how TS handles Decorators when `useDefineForClassFields` is disabled.
My PR also keeps support for the current `makeObservable` approach which _does_ have an expectation of `useDefineForClassFields: true`, so my _guess_ is that you could set it to `false`, but you'd have to entirely use decorators and make sure not to use `makeobservable`.
But all that's a mildly-educated guess / "in theory", so apply salt liberally.

**Reply by belaczek** [reply]
Thanks for your answer! 
> guess is that you could set it to false, but you'd have to entirely use decorators and make sure not to use `makeobservable`

This would be great if true, since it would basically become backward compatible with how mobx 5/4 works. Looking at [download numbers](https://www.npmjs.com/package/mobx?activeTab=versions), there still seem to be quite many projects stuck on older mobx versions because they're probably unable to migrate to `useDefineForClassFields: true`, including us.

**Reply by kubk** [reply]
@belaczek `useDefineForClassFields` makes TypeScript's property transpilation match ECMAScript's behavior. If you specify `compilerOptions.target` as ESNext or ES2022 you already get `useDefineForClassFields` turned on by default. It is something that people have to migrate eventually:
- https://angular.schule/blog/2022-11-use-define-for-class-fields
- https://www.typescriptlang.org/tsconfig#useDefineForClassFields

**Reply by Matchlighter** [reply]
Yeah, I'm one of those numbers, more though b/c I couldn't give up decorators! :P

I can't recall exactly what the issue with `useDefineForClassFields` was and what all it impacts, but from what I can tell, the main issue as it affects MobX is `makeAutoObservable()` and it would overwrite `defineProperty` calls made by MobX 4/5 and 3rd-party decorators. (See also https://github.com/mobxjs/mobx/discussions/3216).

The `accessor` keyword seems to fix the latter issue for both `legacy` and Stage 3 Decs.

Looking at the TS playground, neither Stage 3 nor legacy decorator implementation are directly affected by toggling `useDefineForClassFields`.
https://www.typescriptlang.org/play?strict=false&noImplicitAny=false&strictNullChecks=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&alwaysStrict=false&emitDecoratorMetadata=false&useDefineForClassFields=true#code/FAYwNghgzlAEAqBTKAXWBvYtuwAIgHsBbABwFcVEATWAc0RQAoBKDLHDgJwbM4DtYAVgDc7bAF8xeAgCMoiTgDcIMsIlgQQIZFAKdYKABYBLPrVgBeIaI65Z8pSrUatOvQZNmATMCl25Csqq6hBGpuZWIn72gU4hYd6ikkA

As kubk noted, transitioning to `useDefineForClassFields` is definitely ideal and will likely be the most future proof, but getting there is certainly a pain to say the least.

TL;DR, I _think_ Stage 3 Decorators and my implementation of them in MobX will work without `useDefineForClassFields`.

**Reply by TechQuery** [reply]
@Matchlighter This is Proposal issues of Class fields: tc39/proposal-class-fields#329

**Comment by dartess** [comment]
@Matchlighter Sorry for the naive question, but I don't understand something.
If we have `accessors`, do we really still need `@action`? I'm not familiar with the internals of mobx, but when I saw the accessor decorators, I immediately thought that it would be cool to throw out thousands of dumb actions, because now it is possible to do something like `runInAction` in the accessor decorator setter and just use `this.field = newValue`. I'm naive and it's impossible?

**Reply by Matchlighter** [reply]
Put that way, `@action` was never a _requirement_. The primary purpose of `@action` is so MobX can batch changes. For example, if I were to write:
```
this.x = 5;
this.y = 5;
```
and I had some component that rendered `this.x + this.y`, the component would end up rendering twice - once when `x` changes and once when `y` changes. But if I instead wrap that in an action (either via `runInAction` or via `@action`):
```
runInAction(() => {
  this.x = 5;
  this.y = 5;
})
```
the component would only render once - when the action is over. This is true of actions calling other actions as well - the component will only render once the outer-most action closes.

MobX _does_ have a config option that allows you to change observables outside of actions, but such is viewed as bad practice since it increases risk of needless rendering.

**Comment by mweststrate** [comment]
Just merged :)

**Comment by gilamran** [comment]
Great, but I must say that I'm a bit confused on how my configuration should look like in order to use decorators on my MobX code

- What's the minimal TypeScript version that I must use?
- Should I use `experimentalDecorators: true` on my `tsconfig.json` ?
- Same question regarding `useDefineForClassFields`
- Do I still have to call `makeObservable` on my constructors?

Thanks!

**Comment by mweststrate** [comment]
See https://mobx.js.org/enabling-decorators.html

**Comment by dartess** [comment]
I still don't understand one thing.

First the docs says:

> With modern decorators, it is no longer needed to call makeObservable / makeAutoObservable.

Then the docs says:

> @action some_field = () => {} was and is valid usage (if makeObservable() is also used). However, @action accessor some_field = () => {} is never valid.

So do I still need `makeObservable` or not if I have actions in my code? Does this only apply to actions declared as properties? What about actions declared as methods, but decorated with `action.bound`?

I believe that this piece is completely unclear and requires rewriting.

**Reply by clins1994** [reply]
Confusing indeed

**Reply by kiejo** [reply]
It looks like this got addressed in [mobx@6.13.3](https://github.com/mobxjs/mobx/releases/tag/mobx%406.13.3): `Fix 2022.3 @action decorators on fields no longer require makeObservable`.

So starting with that version, there should be no need to use `makeObservable` with modern decorators anymore.
